client/benchmark/src/main/java/org/elasticsearch/client/benchmark/AbstractBenchmark.java

```diff
@@ -0,0 +1,88 @@
+import org.elasticsearch.client.benchmark.ops.bulk.BulkRequestExecutor;
+import org.elasticsearch.client.benchmark.ops.search.SearchBenchmarkTask;
+import org.elasticsearch.client.benchmark.ops.search.SearchRequestExecutor;
+import org.elasticsearch.common.SuppressForbidden;
+
+import java.io.Closeable;
+
+public abstract class AbstractBenchmark<T extends Closeable> {
+    private static final int SEARCH_BENCHMARK_ITERATIONS = 10_000;
+
+    protected abstract T client(String benchmarkTargetHost) throws Exception;
+
+    protected abstract BulkRequestExecutor bulkRequestExecutor(T client, String indexName, String typeName);
+
+    protected abstract SearchRequestExecutor searchRequestExecutor(T client, String indexName);
+
+    @SuppressForbidden(reason = "system out is ok for a command line tool")
+    public final void run(String[] args) throws Exception {
+        if (args.length < 6) {
+            System.err.println(
+                "usage: benchmarkTargetHostIp indexFilePath indexName typeName numberOfDocuments bulkSize [search request body]");
+            System.exit(1);
+        }
+        String benchmarkTargetHost = args[0];
+        String indexFilePath = args[1];
+        String indexName = args[2];
+        String typeName = args[3];
+        int totalDocs = Integer.valueOf(args[4]);
+        int bulkSize = Integer.valueOf(args[5]);
+
+        int totalIterationCount = (int) Math.floor(totalDocs / bulkSize);
+        // consider 40% of all iterations as warmup iterations
+        int warmupIterations = (int) (0.4d * totalIterationCount);
+        int iterations = totalIterationCount - warmupIterations;
+        String searchBody = (args.length == 7) ? args[6] : null;
+
+        T client = client(benchmarkTargetHost);
+
+        BenchmarkRunner benchmark = new BenchmarkRunner(warmupIterations, iterations,
+            new BulkBenchmarkTask(
+                bulkRequestExecutor(client, indexName, typeName), indexFilePath, warmupIterations + iterations, bulkSize));
+
+        try {
+            benchmark.run();
+            if (searchBody != null) {
+                for (int run = 1; run <= 5; run++) {
+                    System.out.println("=============");
+                    System.out.println(" Trial run " + run);
+                    System.out.println("=============");
+
+                    for (int throughput = 100; throughput <= 100_000; throughput *= 10) {
+                        //force a GC between trials
+                        System.gc();
+
+                        BenchmarkRunner searchBenchmark = new BenchmarkRunner(SEARCH_BENCHMARK_ITERATIONS, SEARCH_BENCHMARK_ITERATIONS,
+                            new SearchBenchmarkTask(
+                                searchRequestExecutor(client, indexName), searchBody, 2 * SEARCH_BENCHMARK_ITERATIONS, throughput));
+                        System.out.printf("Target throughput = %d ops / s%n", throughput);
```

Comment:
I must admit that this terminology is not that clear to me. That may just be because I am not super familiar with benchmarking terminology... I do understand what throughput means, but here we just send one request, then wait a bit, then send another one and so on till we did enough iterations, and the less we wait the higher the targeted throughput, right? I am not sure how we should compare targeted throughput with the obtained one?